// @aws-cdk/aws-lambda-nodejs/lib/package-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageManager = exports.LockFile = void 0;
const os = require("os");
const path = require("path");
const types_1 = require("./types");
var LockFile;
(function (LockFile) {
    LockFile["NPM"] = "package-lock.json";
    LockFile["YARN"] = "yarn.lock";
    LockFile["PNPM"] = "pnpm-lock.yaml";
})(LockFile = exports.LockFile || (exports.LockFile = {}));
/**
 * A node package manager
 */
class PackageManager {
    constructor(props) {
        this.lockFile = props.lockFile;
        this.installCommand = props.installCommand;
        this.runCommand = props.runCommand;
        this.argsSeparator = props.argsSeparator;
    }
    /**
     * Use a lock file path to determine the package manager to use. Optionally, specify a log level to
     * control its verbosity.
     * @param lockFilePath Path of the lock file
     * @param logLevel optional log level @default LogLevel.INFO
     * @returns the right PackageManager for that lock file
     */
    static fromLockFile(lockFilePath, logLevel) {
        const lockFile = path.basename(lockFilePath);
        switch (lockFile) {
            case LockFile.YARN:
                return new PackageManager({
                    lockFile: LockFile.YARN,
                    installCommand: logLevel && logLevel !== types_1.LogLevel.INFO ? ['yarn', 'install', '--no-immutable', '--silent'] : ['yarn', 'install', '--no-immutable'],
                    runCommand: ['yarn', 'run'],
                });
            case LockFile.PNPM:
                return new PackageManager({
                    lockFile: LockFile.PNPM,
                    installCommand: logLevel && logLevel !== types_1.LogLevel.INFO ? ['pnpm', 'install', '--reporter', 'silent'] : ['pnpm', 'install'],
                    runCommand: ['pnpm', 'exec'],
                    argsSeparator: '--',
                });
            default:
                return new PackageManager({
                    lockFile: LockFile.NPM,
                    installCommand: logLevel ? ['npm', 'ci', '--loglevel', logLevel] : ['npm', 'ci'],
                    runCommand: ['npx', '--no-install'],
                });
        }
    }
    runBinCommand(bin) {
        const [runCommand, ...runArgs] = this.runCommand;
        return [
            os.platform() === 'win32' ? `${runCommand}.cmd` : runCommand,
            ...runArgs,
            ...(this.argsSeparator ? [this.argsSeparator] : []),
            bin,
        ].join(' ');
    }
}
exports.PackageManager = PackageManager;
